### PR TITLE
ssm connection: pull bucket region info rather than taking from region var

### PR DESCRIPTION
**Pull request by ljblakey**
##### SUMMARY
Fix issue where syntax error is reported if using ssm connection and the target node is located in a different region to the s3 bucket. 

Fixes https://github.com/ansible-collections/community.aws/issues/1190, https://github.com/ansible-collections/community.aws/issues/637

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
aws_ssm connection plugin

##### ADDITIONAL INFORMATION
When using SSM for ansible connection and the target node is in a different region to the s3 bucket used, the playbook immediately errors with the following. (There are no issues when both target and s3 bucket are in the same region)

`fatal: [i-04444a7f03cc2bffd]: FAILED! => {"ansible_facts": {}, "changed": false, "failed_modules": {"ansible.legacy.setup": {"failed": true, "module_stderr": "", "module_stdout": "  File \"/tmp/ansible/ansible-tmp-1653576081.8378458-29658-258097978113216/AnsiballZ_setup.py\", line 1\r\r\n    <?xml version=\"1.0\" encoding=\"UTF-8\"?>\r\r\n    ^\r\r\nSyntaxError: invalid syntax\r\r", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}}, "msg": "The following modules failed to execute: ansible.legacy.setup\n"}`

The tmp file has the following contents

`<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AuthorizationQueryParametersError</Code><Message>Error parsing the X-Amz-Credential parameter; the region 'us-east-1' is wrong; expecting 'eu-west-1'</Message><Region>eu-west-1</Region><RequestId>4VTAGR4C1V9ATBJT</RequestId><HostId>OahjGsFQHlr3ihxobH/yyH7Mzxq98mwjcb6+J3Y2EifDU7FykCe8b6QJTNodIG5WSquVeJF+Zsk=</HostId></Error>`

Steps to reproduce: run an ansible playbook using the following:
- aws_ec2 inventory
- aws_ssm connection type
- specify an s3 bucket (ansible_aws_ssm_bucket_name var) that is in a different location to the target node

The presigned url generated includes the region the s3 bucket is in, so this region must be used for the session obtained in the `_get_url` function.

### CODE REVIEW

File: plugins/connection/aws_ssm.py [hunk 1]
```diff
@@ -534,7 +534,13 @@ def _flush_stderr(self, subprocess):
 
     def _get_url(self, client_method, bucket_name, out_path, http_method, profile_name, extra_args=None):
         ''' Generate URL for get_object / put_object '''
-        region_name = self.get_option('region') or 'us-east-1'
+
+        client_s3 = boto3.client('s3')
+        bucket_location = client_s3.get_bucket_location(
+            Bucket=(self.get_option('bucket_name')),
+        )
+        region_name = bucket_location['LocationConstraint']
```
Comment (charles-paul-mox) [inline plugins/connection/aws_ssm.py:542]: Looks good - but I believe the addressing style needs to be hardcoded to virtual to work with regioned buckets. See https://docs.aws.amazon.com/cli/latest/topic/s3-config.html#addressing-style and https://github.com/ansible-collections/community.aws/pull/743/files

**Review by charles-paul-mox — CHANGES_REQUESTED** [review]

File: plugins/connection/aws_ssm.py (hunk 1 above)
Comment (charles-paul-mox) [reply plugins/connection/aws_ssm.py:542]: Thanks @lblakey - Corporate IP policies means I cannot merge PRs unfort.

**Review by tremble — APPROVED** [review]

File: changelogs/fragments/1176-ssm-connection-plugin-region-fix.yml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
```
Comment (markuman) [inline changelogs/fragments/1176-ssm-connection-plugin-region-fix.yml:1]:
```suggestion
bugfixes:
```

I wonder if we should flag this as a bugfix?

**Comment by phene** [comment]
FYI @JLukeBlakey -- See #1428

**Comment by charles-paul-mox** [comment]
Hi @phene  - please see this (closed) PR as well: https://github.com/ansible-collections/community.aws/pull/743/files

**Comment by demahum** [comment]
Changes here introduced the usage of `get_bucket_location`.

Is that call too conservative in this case considering that:

> To use this implementation of the operation, you must be the bucket owner.

https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketLocation.html

**Comment by Hokwang** [comment]
This should be roll-back. 
1. get-bucket-location has a problem when bucket is in different region with machine.

2. and aws recommend use `headbucket`
https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketLocation.html

**Comment by tremble** [comment]
@Hokwang ,

Thank you for taking the time to look into this.  However, adding a comment to PR that was merged over a year ago generally isn't the best way to raise issues.

> get-bucket-location has a problem when bucket is in different region with machine.

If you're seeing a specific issue then please open an Issue and be more specific about what you believe is broken (i.e. error messages, which version of the code you're using, etc).  "It has a problem" does nothing to aid in the diagnosis of the issue and if we can't reproduce the issue, we can't test fixes.

> and aws recommend use headbucket

This change predated amazon's recommendation by about 9 months.  You're welcome to open a PR if you believe you have a better way to handle this.

**Pull request merged**